### PR TITLE
Addition DVID and CloudVolume Remotes

### PR DESCRIPTION
**Pull request by lrodri29**
Addition of DVID and Cloud volume remotes. 

All these changes were reviewed and edited in the aplbrain/intern repo, this PR's purpose is to bring in those reviewed updates. 

Relevant Links: 
[DVIDRemote PR](https://github.com/aplbrain/intern/pull/3)
[CloudVolumeRemote PR](https://github.com/aplbrain/intern/pull/4)

### CODE REVIEW

**Comment by j6k4m8** [comment]
We should be sure to "Squash and Merge" when we merge this in, to keep the git history succinct and meaningful.

**Comment by movestill** [comment]
> We should be sure to "Squash and Merge" when we merge this in, to keep the git history succinct and meaningful.

Definitely agree!

**Comment by sandyhider** [comment]
It looks like the example files don't have our copyright notice on them.  Is that done on purpose because they are examples?

**Comment by lrodri29** [comment]
This is missing updates to the CHANGELOG and I think I want to do what @movestill recommended as part of the PR. It's pretty smart

File: examples/dvid/general_test.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+# Create the cutout
+dvid.create_cutout(instance_setup_up, 0, [0, 454], [0, 480], [0, 1], data_tile)
+print("Create cutout successful")
+
+# Get the cutout
+got_cutout = dvid.get_cutout(instance_setup_up, 0, [0, 454], [0, 480], [0, 1])
+
+# Check for equality
+if (got_cutout == data_tile).all():
```
Comment (movestill) [inline examples/dvid/general_test.py:116]: Is there some fancy overload of `==` for dvid cutouts that returns something other than a bool?

File: intern/remote/cv/remote.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+    def get_info(self, resource):
+        """
+        Returns a JSON of the resource properties.
+
+        Args:
+            resource (CloudVolumeResource object)
+
+        Returns:
+            string: JSON format of properties
```
Comment (movestill) [inline intern/remote/cv/remote.py:151]: This should be `(str)`. Likewise for methods below.

File: intern/remote/cv/tests/test_local_remote.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
+import os
```
Comment (movestill) [inline intern/remote/cv/tests/test_local_remote.py:1]: This one needs copyright.

File: intern/resource/dvid/tests/test_data_instance.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        #Set the datatype now
+        data_instance.datatype = 'uint16'
+
+        # Now we should be ready
+        self.assertTrue(data_instance.cutout_ready)
+
+    def test_valid_volume(self):
+        self.assertTrue(self.data_instance_master.valid_volume())
+
+    def test_validate_datatype_uint8(self):
```
Comment (movestill) [inline intern/resource/dvid/tests/test_data_instance.py:61]: I didn't know about `subTest()` when I did the same thing for Boss. It'd save a lot of typing.

File: intern/service/cv/metadata.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        """
+        return resource.cloudvolume.layer
+
+    def set_layer(self, resource, layer):
+        """
+        Set a new layer and commits it to the info file.
+
+        Args:
+            resource (CloudVolume Resource Object)
+            layer (string)
```
Comment (movestill) [inline intern/service/cv/metadata.py:142]: Another string slipped through.

File: examples/dvid/general_test.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+print(dvid.get_info(instance_setup_merge))
+print(dvid.get_server_info())
+print(dvid.get_server_compiled_types())
+dvid.server_reload_metadata()
+
+########### Test Voluming API ###########
+#
+
+# Prepare the data
+img = Image.open("/Users/rodrilm2/Desktop/humans/0.png")
```
Comment (j6k4m8) [inline examples/dvid/general_test.py:91]: should remove local references

File: intern/remote/cv/remote.py [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+from intern.service.cv.metadata import MetadataService
+from intern.service.cv.project import ProjectService
+
+
+class CloudVolumeRemote(Remote):
+    def __init__(
+        self, cfg_file_or_dict=None,
+    ):
+        """Constructor.
+        Protocol and host specifications are taken in as keys -values of dictionary.
```
Comment (j6k4m8) [inline intern/remote/cv/remote.py:28]: We should include what arguments we expect to receive here in the docstring

File: intern/service/dvid/versioning.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+        """
+        DVIDService.__init__(self)
+        self.base_url = base_url
+
+    def merge(self, UUID, parents, mergeType, note):
+        """ Creates a conflict-free merge of a set of committed parent UUIDs into a child.  Note
+            the merge will not necessarily create an error immediately
+
+        Args:
+            mergeType (str) = "conflict-free"
```
Comment (movestill) [inline intern/service/dvid/versioning.py:43]: Is this like a commit message? Assume this is something a dvid user would know what's valid to provide?

File: intern/remote/cv/remote.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+        factor=(2, 2, 1),
+    ):
+        """
+        Creates the info JSON necessary for a new cloudvolume resource.
+        Required:
+            num_channels: (int) 1 for grayscale, 3 for RGB
+            layer_type: (str) typically "image" or "segmentation"
+            data_type: (str) e.g. "uint8", "uint16", "uint32", "float32"
+            resolution: int (x,y,z), x,y,z voxel dimensions in nanometers
+
```
Comment (j6k4m8) [inline intern/remote/cv/remote.py:82]: extra linefeed

File: intern/remote/dvid/remote.py [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+			datatype (str): data type of the instance (uint8, uint16, uint64) defaults to uint8
+		Returns:
+			resource (intern.resource.dvid.DvidResource)
+		"""
+        return DataInstanceResource(data_instance, UUID, datatype=datatype)
+
+    def get_cutout(self, resource, res, xrange, yrange, zrange, **kwargs):
+        """Method to request a volume of data from DVID server
+
+		Args:
```
Comment (j6k4m8) [inline intern/remote/dvid/remote.py:183]: low priority, we can hit later. but these docstrings should have 4 ~8~ space docstrings

File: requirements.txt [hunk 11]
```diff
@@ -4,3 +4,7 @@ blosc>=1.4.4
 six
 mock
 nose2
+
+# cloud-volume
```
Comment (movestill) [inline requirements.txt:8]: Can remove this.

File: intern/service/dvid/versioning.py (hunk 8 above)
Comment (lrodri29) [reply intern/service/dvid/versioning.py:43]: right

File: intern/resource/dvid/tests/test_data_instance.py (hunk 4 above)
Comment (lrodri29) [reply intern/resource/dvid/tests/test_data_instance.py:61]: haha nice, I did the same things here. I think it's fine though

File: requirements.txt (hunk 11 above)
Comment (j6k4m8) [reply requirements.txt:8]: sorry, my bad — this was to document that the below are for cv installs only. (we'll eventujally want to separate this out as `pip install intern[cloudvolume]` or something, so i figured i'd keep them documented.

File: examples/dvid/general_test.py (hunk 1 above)
Comment (j6k4m8) [reply examples/dvid/general_test.py:116]: `nparray(A) == nparray(B)` returns a nparray binary mask of the same size, `.all()` converts that to a single bool if all values are True

File: intern/remote/dvid/remote.py (hunk 10 above)
Comment (lrodri29) [reply intern/remote/dvid/remote.py:183]: 👍

File: intern/remote/cv/remote.py (hunk 2 above)
Comment (lrodri29) [reply intern/remote/cv/remote.py:151]: Done

File: intern/remote/cv/tests/test_local_remote.py (hunk 3 above)
Comment (lrodri29) [reply intern/remote/cv/tests/test_local_remote.py:1]: Done

File: intern/service/cv/metadata.py (hunk 5 above)
Comment (lrodri29) [reply intern/service/cv/metadata.py:142]: done

File: intern/remote/cv/remote.py (hunk 9 above)
Comment (lrodri29) [reply intern/remote/cv/remote.py:82]: Done

File: intern/remote/cv/remote.py (hunk 7 above)
Comment (lrodri29) [reply intern/remote/cv/remote.py:28]: Done

File: requirements.txt (hunk 11 above)
Comment (movestill) [reply requirements.txt:8]: Oh gotcha. Didn't realize that this indicates the stuff below if cv specific. Thought it was a repeat of the cv requirement.

**Pull request merged**